### ISSUE TITLE
Dark mode support via @media (prefers-color-scheme: dark)

### ISSUE THREAD

**Issue by lutzroeder** [electron/electron#15540]
Safari Technology Preview now supports dark mode via `@media (prefers-color-scheme: dark)` in CSS media queries. 

Same should be supported in Electron apps to enable automatic light/dark mode switching.

**Comment by MarshallOfSound** [comment]
Electron has dark mode API's through `systemPreferences.isDarkMode()`.  Any CSS features should be requested through chromium.

**Comment by lutzroeder** [comment]
Chromium [#889087](https://bugs.chromium.org/p/chromium/issues/detail?id=889087) for reference.

**Comment by caesar** [comment]
Chromium supports `@media (prefers-color-scheme: dark)` now in version 76, but Electron 6.0.0-beta4 still doesn't, even though it's based on Chrome 76.0.3783.1.

Maybe this should be reopened?

**Comment by MarshallOfSound** [comment]
@caesar Once Electron includes a Chromium version where that query is enabled by default it should be picked up without us doing anything.  76 is very much a moving target, it's possible this change will come in on the next roll --> https://github.com/electron/electron/pull/18439

**Comment by caesar** [comment]
Hmm, this still isn't in `electron@6.0.0-beta.8`. It's been in Chromium releases for a while now (including the Chrome Canary and Edge Canary builds). Is there something more to this?

<img width="800" alt="Screenshot 2019-06-18 at 11 35 35" src="https://user-images.githubusercontent.com/1566516/59676835-607c0680-91c0-11e9-9a05-4491517c42bd.png">

**Comment by MarshallOfSound** [comment]
Yup, there's something more to this, missing bit of wiring on our side for web prefs 👍

**Comment by caesar** [comment]
Aha! Good to know it's not just me 😆

**Comment by GitFiend** [comment]
Am I missing something? It seems broken in electron 6.0:

<img width="696" alt="Screen Shot 2019-08-03 at 4 31 37 PM" src="https://user-images.githubusercontent.com/47970508/62407404-3cf40a80-b60c-11e9-99c9-86922eacd32a.png">

**Comment by MarshallOfSound** [comment]
@GitFiend We have an entire guide for this 👍 

https://electronjs.org/docs/tutorial/mojave-dark-mode-guide#mojave-dark-mode

The important part being

> In order for the isDarkMode and Tray APIs to work correctly you need to either have NSRequiresAquaSystemAppearance set to false in your Info.plist file or be on Electron >=7.0.0.

**Comment by lutzroeder** [comment]
@MarshallOfSound If `isDarkMode` returns `true`, the `prefers-color-schema: dark` styles expected to work as well -- correct?

**Comment by MarshallOfSound** [comment]
@lutzroeder Unfortunately _not_. Maybe that guide should be updated to include this information.

`isDarkMode()` is implemented by Electron using API (a) and the `prefers-color-scheme: dark` is implemented by Chromium using API (b).

API (a) is designed to work on all macOS versions and only requires `NSRequiresAquaSystemAppearance=false` on Catalina.

API (b) is designed to work on >= Mojave and requires `NSRequiresAquaSystemAppearance=false` on all supported versions.

TLDR: They are backed by different APIs, the media query requires `NSRequiresAquaSystemAppearance=false`

**Comment by caesar** [comment]
@MarshallOfSound that's very useful info, which should definitely be in the docs as it's rather unintuitive until you understand why.

**Comment by bkawk** [comment]
@media (prefers-color-scheme: dark) does not work on Electron ^6.0.2

**Comment by GitFiend** [comment]
> @media (prefers-color-scheme: dark) does not work on Electron ^6.0.2

It works for me on Windows but not Mac. IsDarkMode works on Mac, but not Windows. I think neither work on Linux, but have only tested Ubuntu.

**Comment by fabiospampinato** [comment]
> TLDR: They are backed by different APIs, the media query requires NSRequiresAquaSystemAppearance=false

And how does one add that key? I'm using `electron-builder` and it offers some options for this, but what if I want to also enabled this during development?

**Comment by MarshallOfSound** [comment]
@fabiospampinato builder has `darkModeSupport` --> https://www.electron.build/configuration/mac.html

**Comment by fabiospampinato** [comment]
@MarshallOfSound Yes but I also want to have this working during development, before using the builder.

**Comment by fabiospampinato** [comment]
@MarshallOfSound This is still broken for me in v6.0.3, the first time I toggle dark mode it's detected, but then it doesn't detect it anymore.

I use the following to detect dark mode: `window.matchMedia ( '(prefers-color-scheme: dark)' )` and this happens:

1. Dark mode disabled => no dark mode detected
2. I enabled dark mode => dark mode detected
3. I disabled dark mode => dark mode still detected

I should I open a separate issue?

**Comment by MarshallOfSound** [comment]
> I should I open a separate issue?

If you have a `gist` that reproduces this yes, but please triple check.  I just followed your steps above and it worked perfectly for me.

**Comment by fabiospampinato** [comment]
I quadruple checked, and it always works the same for me. Additionally if dark mode is enabled when starting the app then it's always detected, basically Electron/Chromium gets stuck once dark mode is enabled.

I don't have a repro project, but since it works for you maybe the issue is that I'm on v10.14.6 or something? This is my `Info.plist`:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>BuildMachineOSBuild</key>
    <string>17D102</string>
    <key>CFBundleDisplayName</key>
    <string>Electron</string>
    <key>CFBundleExecutable</key>
    <string>Electron</string>
    <key>CFBundleIconFile</key>
    <string>electron.icns</string>
    <key>CFBundleIdentifier</key>
    <string>com.github.Electron</string>
    <key>CFBundleInfoDictionaryVersion</key>
    <string>6.0</string>
    <key>CFBundleName</key>
    <string>Electron</string>
    <key>CFBundlePackageType</key>
    <string>APPL</string>
    <key>CFBundleShortVersionString</key>
    <string>6.0.3</string>
    <key>CFBundleVersion</key>
    <string>6.0.3</string>
    <key>DTCompiler</key>
    <string>com.apple.compilers.llvm.clang.1_0</string>
    <key>DTSDKBuild</key>
    <string>10.13</string>
    <key>DTSDKName</key>
    <string>macosx10.13</string>
    <key>DTXcode</key>
    <string>0941</string>
    <key>DTXcodeBuild</key>
    <string>9F2000</string>
    <key>LSApplicationCategoryType</key>
    <string>public.app-category.developer-tools</string>
    <key>LSMinimumSystemVersion</key>
    <string>10.10.0</string>
    <key>NSHighResolutionCapable</key>
    <true/>
    <key>NSMainNibFile</key>
    <string>MainMenu</string>
    <key>NSPrincipalClass</key>
    <string>AtomApplication</string>
    <key>NSSupportsAutomaticGraphicsSwitching</key>
    <true/>
    <key>NSRequiresAquaSystemAppearance</key>
    <false/>
  </dict>
</plist>
```

**Comment by johannesjo** [comment]
Should we open up a new issue for this one or can we maybe reopen it?

**Comment by konsumer** [comment]
I think one important thing to note about this, is that dark-mode isn't just for mac, as chromium supports it for Linux & Windows, at this point. I use the CSS selectors to pick a different theme on the web, which works for stable chromium:

```html
<link rel="stylesheet" href="dark.css" media="(prefers-color-scheme: dark)">
<link rel="stylesheet" href="light.css" media="(prefers-color-scheme: no-preference), (prefers-color-scheme: light)">
```

The non-standard electron js API makes this a lot more complicated to use.

Here is Chromium (on top) responding to theme-change on linux. Electron is on bnottom, and it doesn't respond, with the same web-page.
![Peek 2019-12-06 18-55](https://user-images.githubusercontent.com/83857/70367931-1fed7380-185a-11ea-8044-840920e3bfcf.gif)

**Comment by konsumer** [comment]
Another thing about this is that electron is actively matching to `prefers-color-scheme: light`, on a dark system. If I remove the light link line, above, it has no style (inverted-match on `prefers-color-scheme: dark`, so it doesn't load the dark line.) This definitely seems like a bug.

**Comment by MarshallOfSound** [comment]
@konsumer use Electron 7.

**Comment by konsumer** [comment]
@MarshallOfSound I am using electron 7.1.3.

**Comment by MarshallOfSound** [comment]
@konsumer Electron doesn't implement a non-standard API.  We provide the exact same API backed by the exact same logic that Chrome provides (the CSS media query).

We also expose an **additional** API surface so that you can query and control the API from the main process.  If something is working in Chromium 78 and not working in Electron 7 with regards to the CSS media query.  Please raise an issue with repro steps.

**Comment by konsumer** [comment]
@MarshallOfSound What I am saying is the only way to respond to theme changes is an API that isn't available in other browsers (as you say "an additional API surface") but since the regular way doesn't work, in practice that means "we support it with a non-standard API, not the regular way, with media-queries"

**Comment by MarshallOfSound** [comment]
@konsumer You don't understand what I'm saying.

The media query **does** work.  I just double checked on my macOS and windows machine just now.  I don't have a linux VM handy at home so can't verify on that.  But the media query **does** work.  The `nativeTheme` module is just **additional** API surface, the media query is 100% supported.

**Comment by konsumer** [comment]
Hmm, the screenshot you are seeing above is on linux, with electron 7.1.3. Do I need to do something different?

**Comment by MarshallOfSound** [comment]
🤷‍♂As I said, if you have a repro that shows the media query not working in the same way that it works in Chrome.  Raise a new issue with that repro included and exact instructions for OS, theme selection, etc.

**Comment by konsumer** [comment]
I see what you are saying. It's not *this* issue, and I should raise a new one. Will do.

**Comment by konsumer** [comment]
Filed #21427
